packages/http/src/router/index.ts

```diff
@@ -23,26 +22,25 @@ export const router: IRouter<IHttpOperation, IHttpRequest, IHttpConfig> = {
       throw NO_RESOURCE_PROVIDED_ERROR;
     }
 
-    let noServerProvided: boolean = true;
     let noneMethodMatched: boolean = true;
     let nonePathMatched: boolean = true;
-    let noneServerMatched: boolean = !ignoreServers;
+    let noneServerMatched: boolean = true;
 
     for (const resource of resources) {
       if (!matchByMethod(input, resource)) continue;
       noneMethodMatched = false;
 
       const pathMatch = matchPath(requestPath, resource.path);
-      const { servers = [] } = resource;
-
-      if (servers.length) noServerProvided = false;
       if (pathMatch !== MatchType.NOMATCH) nonePathMatched = false;
 
+      const { servers = [] } = resource;
       let serverMatch: MatchType | null = null;
 
-      if (!ignoreServers) {
+      if (!ignoreServers && servers.length > 0) {
```

Comment:
What if we **don't** want to ignore servers AND no servers is found?

You are missing a logical branch here.

```
if(!ignoreServers) {
  if (servers.length > 0) {
  } else {
    // this is missing
  }
}

Reply:
In case described below we should still throw the error you have removed

Reply:
All branches are covered here. The code you wrote is not equivalent of the above.

`if (A && B) 1; else 2;`  is a shorthand of  `if A { if B 1; else 2; } else { if B 2; else 2; }`

Reply:
Ok, I see but I'm saying we need the following logic (I overdid it on purpose):

```
if(!ignoreServers) {
  if (servers.length > 0) {
    // ...
    noServerProvided = false;
    noneServerMatched = false;
  } else {
    // Next line is important. We have explicitly asked to match servers but no server was provided. This code removes this functionality.
    noServerProvided = true;
    noneServerMatched = true;
  }
} else {
   noServerProvided = false;
   noneServerMatched = false;
}
```

How do you want to achieve the same logic with two branches?

```
if(!ignoreServers && servers.length > 0) {
   // ...
} else {
   noneServerMatched = false;
}
```

Reply:
This is the point of this change. If none server provided it should still work (see the jira issue). That is why I removed noServerProvided errors. That's not relevant anymore.

Btw. see that line: https://github.com/stoplightio/prism/pull/240/files#diff-f9a10b37616fb5669ecd5218fc8535c9R27

Reply:
> This is the point of this change. If none server provided it should still work (see the jira issue). That is why I removed noServerProvided errors. That's not relevant anymore.
> 
> Btw. see that line: https://github.com/stoplightio/prism/pull/240/files#diff-f9a10b37616fb5669ecd5218fc8535c9R27

If I understand it correctly the purpose of this task is to provide response **if there is no servers (servers = []) AND if we did not ask to validate servers (ignoreServers=true)**.

This is different from the case I'm making. I'm telling **I explicitly ask you to validate server (ignoreServers=false) AND there is no servers provided (servers = [])**.

The two are different.

Reply:
Definitely very confusing. 

```
 const ignoreServers: boolean = requestBaseUrl === undefined;
```

I thought we completely ignored base paths, and I'm not sure why we are turning ignoreServers on or off based on wether or not a base path exists? Base path is not even a concept in OpenAPI v3.

Do we have server validation? Like, if i pass in ?__server=foo.com it should error if foo.com does not exist in the spec, right? 

- if (user specified a server - server validation is on)
  - ❌ no servers exist, so fail (cant validate against nothing)
  - ❌ servers exist and this is not one of them
  - ✅ servers exist and this is one of them
- else
  - there is no else nobody cares about servers yaaaaay

Am I missing something?

Reply:
Also `noneServerMatched` is a really confusing name for a variable. negative variable names for booleans lead to double negatives that are hard to understand. 

can we have positive variable names? 

- `serverValidationEnabled`
- `serverFound`

that sort of thing.

Reply:
> Definitely very confusing.
> 
> ```
>  const ignoreServers: boolean = requestBaseUrl === undefined;
> ```
> I thought we completely ignored base paths, and I'm not sure why we are turning ignoreServers on or off based on wether or not a base path exists? Base path is not even a concept in OpenAPI v3.

OpenAPI doesn't have base path _per se_ but it has `servers`. Servers in OAS3 are pretty flexible and can be templated btw. `requestBaseUrl` is basically a flag to turn on/off server/basePath validation.


> 
> Do we have server validation? Like, if i pass in ?__server=foo.com it should error if foo.com does not exist in the spec, right?

AFAIK we do or we have partially implemented it (at the framework level). Anyway we **should** have it. 

> 
> * if (user specified a server - server validation is on)
>   
>   * x no servers exist, so fail (cant validate against nothing)
>   * x servers exist and this is not one of them
>   * white_check_mark servers exist and this is one of them
> * else
>   
>   * there is no else nobody cares about servers yaaaaay
> 
> Am I missing something?

Sounds about right!

Reply:
@StefanDywersant this is still relevant btw ^ My first comment is still valid.

Reply:
Check out the README, I put a FAQ section in there: 

> **The API description defines a base path of `/api` (using OpenAPI v2 `basePath` keyword, or in
> OpenAPI v3 using a path in `servers.url`), but requests seem to fail when using it?**
> 
> Base paths are completely ignored by the Prism HTTP server, so they can be removed from the request.
> If you have a base path of `/api` and your path is defined as `hello`, then a request to
`http://localhost:4010/hello` would work, but `http://localhost:4010/api/hello` will fail.

Let me know if that is not true. 

If we do not yet have any server validation then this code is even more confusing :D

Reply:
> `requestBaseUrl` is basically a flag to turn on/off server/basePath validation.

Maybe that variabke could be `serverValidationEnabled` or something that starts with `is, so its a bit more clear :) I dont want to bikeshed over variable naming but this whole thing is confusing to the larger intent, and without docs we have to read the code, and I cant read this code! 😓

Reply:
Summary of changes made:
- introduced `any[Method,Path,Server][Matched,Provided]` for clarity what those vars do
- fixed logic, handling `servers exist and this is not one of them` case
- updated tests